File: android/app/src/main/java/com/buddy/app/Coordination.kt
package com.buddy.app

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Handler
import android.os.Looper
import android.util.Log
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket

// Android side of protocol/coordination.md. v1 scope mirrors the mac:
// mDNS announce/browse (NSD), JSON-lines TCP, epochs, travel both ways.
class Coordination(context: Context, private val onUi: Handler = Handler(Looper.getMainLooper())) {
    companion object {
        // Legacy shared literal - only the migration default. Real installs
        // carry a per-pair secret in prefs ("secret"); the mac side reads its
        // own from ~/.buddy/secret, and re-keying means writing the same
        // value on both sides. Never ship new pairings on the literal.
        const val LEGACY_SECRET = "buddy-doorknob"
        const val SERVICE_TYPE = "_buddy._tcp."
        const val PORT = 47800
        const val HB_PORT = 47801
        const val SILENCE_MS = 15_000L
        const val TAG = "BuddyCoord"

        fun rankOf(deviceId: String): Int = when (deviceId) {
            "mac" -> 1
            "phone" -> 2
            else -> 9
        }
    }

    private val prefs = context.getSharedPreferences("coordination", Context.MODE_PRIVATE)
    private val secret: String get() = prefs.getString("secret", null) ?: LEGACY_SECRET
    var epoch: Int
        get() = prefs.getInt("epoch", 0)
        private set(v) { prefs.edit().putInt("epoch", v).apply() }
    var ownsBuddy: Boolean
        get() = prefs.getBoolean("owner", false)
        private set(v) { prefs.edit().putBoolean("owner", v).apply() }

    var onArrive: ((JSONObject) -> Unit)? = null
    var onDepart: (() -> Unit)? = null
    // Crash election: the mac (the only non-phone peer) owned buddy and went
    // silent past the timeout. Payload = last replicated snapshot.
    var onEmergencyClaim: ((JSONObject) -> Unit)? = null
    // Every replicated state snapshot from the owner (traits sync etc).
    var onSnapshot: ((JSONObject) -> Unit)? = null

    private val appContext = context
    private val nsd = context.getSystemService(Context.NSD_SERVICE) as NsdManager
    private var regListener: NsdManager.RegistrationListener? = null
    // Screen-off wifi filters multicast; without this lock the phone stops
    // answering mDNS the moment the screen sleeps.
    private var mcastLock: android.net.wifi.WifiManager.MulticastLock? = null
    private var server: ServerSocket? = null
    @Volatile private var peerHost: String? = null
    @Volatile private var peerPort: Int = 0
    @Volatile private var running = true

    // Liveness (mesh heartbeats per coordination.md)
    @Volatile private var macLastSeen = 0L
    @Volatile private var macOwns = false
    @Volatile private var macEpoch = 0
    private var lastSnapshot: JSONObject
        get() = try { JSONObject(prefs.getString("snapshot", "{}") ?: "{}") } catch (e: Exception) { JSONObject() }
        set(v) { prefs.edit().putString("snapshot", v.toString()).apply() }

    fun start() {
        prefs.getString("peerHost", null)?.let {
            peerHost = it
            peerPort = prefs.getInt("peerPort", PORT)
        }
        val wifi = appContext.applicationContext
            .getSystemService(Context.WIFI_SERVICE) as android.net.wifi.WifiManager
        mcastLock = wifi.createMulticastLock("buddy-mdns").apply {
            setReferenceCounted(false)
            acquire()
        }
        server = ServerSocket(PORT)
        Thread {
            while (running) {
                try {
                    val sock = server?.accept() ?: break
                    Thread { serve(sock) }.start()
                } catch (e: Exception) {
                    if (running) Log.w(TAG, "accept: $e")
                }
            }
        }.start()

        val info = NsdServiceInfo().apply {
            serviceName = "phone"
            serviceType = SERVICE_TYPE
            port = PORT
        }
        val reg = object : NsdManager.RegistrationListener {
            override fun onServiceRegistered(i: NsdServiceInfo) { Log.i(TAG, "registered ${i.serviceName}") }
            override fun onRegistrationFailed(i: NsdServiceInfo, e: Int) { Log.w(TAG, "register failed $e") }
            override fun onServiceUnregistered(i: NsdServiceInfo) {}
            override fun onUnregistrationFailed(i: NsdServiceInfo, e: Int) {}
        }
        regListener = reg
        nsd.registerService(info, NsdManager.PROTOCOL_DNS_SD, reg)

        nsd.discoverServices(SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, object : NsdManager.DiscoveryListener {
            override fun onServiceFound(s: NsdServiceInfo) {
                // Names are not identities (collision renames, stale records):
                // probe every record and let the hello reply's deviceId decide.
                nsd.resolveService(s, object : NsdManager.ResolveListener {
                    override fun onServiceResolved(r: NsdServiceInfo) {
                        val host = r.host?.hostAddress ?: return
                        probe(host, r.port)
                    }
                    override fun onResolveFailed(i: NsdServiceInfo, e: Int) { Log.w(TAG, "resolve failed $e") }
                })
            }
            override fun onServiceLost(s: NsdServiceInfo) {
                // A lost record's identity is unknowable from its name;
                // peer liveness comes from heartbeat recency instead.
            }
            override fun onDiscoveryStarted(t: String) {}
            override fun onDiscoveryStopped(t: String) {}
            override fun onStartDiscoveryFailed(t: String, e: Int) { Log.w(TAG, "discovery failed $e") }
            override fun onStopDiscoveryFailed(t: String, e: Int) {}
        })

        startHeartbeats()
    }

    fun stop() {
        running = false
        // Withdraw the mDNS record: an announcement lingering after death is
        // exactly what once sent the mac's travel to a dead port.
        regListener?.let { try { nsd.unregisterService(it) } catch (_: Exception) {} }
        regListener = null
        try { mcastLock?.release() } catch (_: Exception) {}
        try { server?.close() } catch (_: Exception) {}
        try { hbSocket?.close() } catch (_: Exception) {}
    }

    // Hello the endpoint; adopt it as the peer only when the reply carries
    // someone else's deviceId - our own records answer as "phone".
    private fun probe(host: String, port: Int) {
        Thread {
            try {
                val sock = Socket()
                sock.connect(InetSocketAddress(host, port), 5000)
                sock.soTimeout = 5000
                send(sock, JSONObject().put("type", "hello").put("epoch", epoch))
                val reader = BufferedReader(InputStreamReader(sock.getInputStream()))
                val line = reader.readLine()
                sock.close()
                val frame = try { JSONObject(line ?: return@Thread) } catch (e: Exception) { return@Thread }
                if (frame.optString("s") != secret) return@Thread
                val from = frame.optString("from")
                if (from.isEmpty() || from == "phone") return@Thread
                peerHost = host
                peerPort = port
                // Last-known address survives process death: heartbeats can
                // resume immediately on restart, before (or without) mDNS.
                prefs.edit().putString("peerHost", host).putInt("peerPort", port).apply()
                BuddyService.peerOnline = true
                Log.i(TAG, "peer $from at $host:$port")
            } catch (e: Exception) {
                Log.w(TAG, "probe $host:$port: $e")
            }
        }.start()
    }

    // MARK: - Heartbeats + crash election

    private var hbSocket: java.net.DatagramSocket? = null

    private fun startHeartbeats() {
        // Listener: track the mac's liveness and whether it owns buddy.
        Thread {
            try {
                val sock = java.net.DatagramSocket(HB_PORT)
                hbSocket = sock
                val buf = ByteArray(4096)
                while (running) {
                    val packet = java.net.DatagramPacket(buf, buf.size)
                    sock.receive(packet)
                    val frame = try {
                        JSONObject(String(packet.data, 0, packet.length))
                    } catch (e: Exception) { continue }
                    if (frame.optString("s") != secret) continue
                    if (frame.optString("from") == "phone") continue
                    macLastSeen = System.currentTimeMillis()
                    macOwns = frame.optBoolean("owner", false)
                    macEpoch = frame.optInt("epoch", 0)
                    BuddyService.peerOnline = true
                }
            } catch (e: Exception) {
                if (running) Log.w(TAG, "hb listen: $e")
            }
        }.start()

        // Sender + election watchdog every 3s.
        Thread {
            while (running) {
                try { Thread.sleep(3000) } catch (e: Exception) { break }
                peerHost?.let { host ->
                    try {
                        val frame = JSONObject()
                            .put("v", 1).put("s", secret).put("from", "phone")
                            .put("type", "heartbeat").put("epoch", epoch)
                            .put("owner", ownsBuddy)
                        val bytes = frame.toString().toByteArray()
                        java.net.DatagramSocket().use {
                            it.send(java.net.DatagramPacket(bytes, bytes.size,
                                java.net.InetAddress.getByName(host), HB_PORT))
                        }
                    } catch (e: Exception) { Log.w(TAG, "hb send: $e") }
                }
                // Silence-based only after the mac has been heard at least
                // once - before that, the probe's verdict stands.
                if (macLastSeen != 0L)
                    BuddyService.peerOnline = System.currentTimeMillis() - macLastSeen < SILENCE_MS
                checkElection()
            }
        }.start()
    }

    // The phone is rank 2 and the mac is rank 1: an election here happens
    // only when the mac owned buddy and crashed (silence past timeout while
    // on the LAN). The phone going silent never triggers anything anywhere -
    // that is the "buddy is out with the human" rule, enforced on the mac side by
    // taking no action on silence at all.
    private fun checkElection() {
        if (ownsBuddy) return
        if (!macOwns) return
        if (macLastSeen == 0L) return
        if (System.currentTimeMillis() - macLastSeen < SILENCE_MS) return
        // Off wifi, WE left the network - the mac didn't die. Election is for
        // a crashed mac observed from inside the same LAN; electing here once
        // resurrected a second buddy while the real one lived on.
        if (!onWifi()) return
        // Heartbeat silence proves nothing while TCP still answers: UDP can
        // die alone (stale IP, VPN routing) with the mac perfectly alive.
        if (macAnswersTcp()) {
            macLastSeen = System.currentTimeMillis()
            return
        }
        val claimed = maxOf(epoch, macEpoch) + 1
        epoch = claimed
        ownsBuddy = true
        macOwns = false
        Log.i(TAG, "emergency claim, epoch $claimed (mac silent)")
        val snapshot = lastSnapshot
        onUi.post { onEmergencyClaim?.invoke(snapshot) }
    }

    private fun onWifi(): Boolean {
        val cm = appContext.getSystemService(Context.CONNECTIVITY_SERVICE)
            as android.net.ConnectivityManager
        val caps = cm.getNetworkCapabilities(cm.activeNetwork) ?: return false
        return caps.hasTransport(android.net.NetworkCapabilities.TRANSPORT_WIFI)
    }

    private fun macAnswersTcp(): Boolean {
        val host = peerHost ?: return false
        return try {
            Socket().use { it.connect(InetSocketAddress(host, peerPort), 4000) }
            true
        } catch (e: Exception) { false }
    }

    val hasPeer: Boolean get() = peerHost != null

    private fun serve(sock: Socket) {
        try {
            val reader = BufferedReader(InputStreamReader(sock.getInputStream()))
            while (true) {
                val line = reader.readLine() ?: break
                val frame = try { JSONObject(line) } catch (e: Exception) { continue }
                if (frame.optString("s") != secret) continue
                val fEpoch = frame.optInt("epoch", -1)
                val type = frame.optString("type")
                if (type != "hello" && fEpoch < epoch) continue

                when (type) {
                    "hello" -> send(sock, JSONObject()
                        .put("type", "state").put("epoch", epoch)
                        .put("seq", 0).put("op", "snapshot").put("owner", ownsBuddy))
                    "travel" -> {
                        val to = frame.optString("to")
                        if (to.isNotEmpty() && to != "phone") {
                            Log.i(TAG, "travel misdelivery (to $to), refused")
                            continue
                        }
                        epoch = fEpoch
                        ownsBuddy = true
                        send(sock, JSONObject().put("type", "travel-ack").put("epoch", epoch))
                        val payload = frame.optJSONObject("payload") ?: JSONObject()
                        Log.i(TAG, "travel in, epoch $epoch")
                        onUi.post { onArrive?.invoke(payload) }
                    }
                    "traitSet" -> {
                        if (ownsBuddy) {
                            val name = frame.optString("name")
                            val value = frame.optDouble("value", 0.5)
                            if (name.isNotEmpty()) onUi.post { onTraitSet?.invoke(name, value) }
                        }
                    }
                    "claim", "state" -> {
                        if (type == "state") {
                            frame.optJSONObject("payload")?.let {
                                lastSnapshot = it
                                onUi.post { onSnapshot?.invoke(it) }
                            }
                        }
                        val theyOwn = if (type == "claim") true else frame.optBoolean("owner", false)
                        val theirRank = rankOf(frame.optString("from"))
                        val theyWin = fEpoch > epoch || (fEpoch == epoch && theirRank < rankOf("phone"))
                        if (theyOwn && theyWin) {
                            epoch = fEpoch
                            if (ownsBuddy) {
                                ownsBuddy = false
                                Log.i(TAG, "ceding ownership to ${frame.optString("from")} epoch $fEpoch")
                                onUi.post { onDepart?.invoke() }
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.w(TAG, "serve: $e")
        } finally {
            try { sock.close() } catch (_: Exception) {}
        }
    }

    private fun send(sock: Socket, frame: JSONObject) {
        frame.put("v", 1).put("s", secret).put("from", "phone")
        sock.getOutputStream().write((frame.toString() + "\n").toByteArray())
        sock.getOutputStream().flush()
    }

    // Ask the owner to change a trait: one soul, owner holds the pen. The
    // owner applies (clamped to the human's bounds) and its state broadcast echoes
    // the result back to every device.
    var onTraitSet: ((String, Double) -> Unit)? = null

    fun sendTraitSet(name: String, value: Double) {
        val host = peerHost ?: return
        val port = peerPort
        Thread {
            try {
                val sock = Socket()
                sock.connect(InetSocketAddress(host, port), 5000)
                send(sock, JSONObject().put("type", "traitSet").put("epoch", epoch)
                    .put("name", name).put("value", value))
                sock.close()
            } catch (e: Exception) {
                Log.w(TAG, "traitSet send: $e")
            }
        }.start()
    }

    // Full-snapshot state event to the peer (spec: replication on every state
    // change). Owner-only, same rule as the mac side.
    fun broadcastState(payload: JSONObject) {
        val host = peerHost ?: return
        val port = peerPort
        if (!ownsBuddy) return
        Thread {
            try {
                val sock = Socket()
                sock.connect(InetSocketAddress(host, port), 5000)
                send(sock, JSONObject().put("type", "state").put("epoch", epoch)
                    .put("seq", 0).put("op", "snapshot").put("owner", true)
                    .put("payload", payload))
                sock.close()
            } catch (e: Exception) {
                Log.w(TAG, "state broadcast: $e")
            }
        }.start()
    }

    // Travel buddy back to the discovered peer. completion on main thread.
    fun travel(payload: JSONObject, completion: (Boolean) -> Unit) {
        val host = peerHost
        val port = peerPort
        if (!ownsBuddy || host == null) {
            onUi.post { completion(false) }
            return
        }
        Thread {
            var ok = false
            val proposed = epoch + 1
            try {
                val sock = Socket()
                sock.connect(InetSocketAddress(host, port), 5000)
                sock.soTimeout = 10000
                send(sock, JSONObject().put("type", "travel").put("epoch", proposed)
                    .put("to", "mac")
                    .put("payload", payload))
                val reader = BufferedReader(InputStreamReader(sock.getInputStream()))
                while (true) {
                    val line = reader.readLine() ?: break
                    val frame = try { JSONObject(line) } catch (e: Exception) { continue }
                    if (frame.optString("s") == secret &&
                        frame.optString("type") == "travel-ack" &&
                        frame.optInt("epoch") == proposed) {
                        ok = true
                        break
                    }
                }
                sock.close()
            } catch (e: Exception) {
                Log.w(TAG, "travel out: $e")
            }
            if (ok) {
                epoch = proposed
                ownsBuddy = false
                Log.i(TAG, "travel out acked, epoch $proposed")
                onUi.post { onDepart?.invoke() }
            }
            onUi.post { completion(ok) }
        }.start()
    }
}

File: macos/Sources/Buddy/Coordination.swift
import Foundation
import Network

// LAN coordination per protocol/coordination.md: mDNS discovery, JSON-lines
// TCP frames, epochs, travel handshake. v1 scope: discovery + travel both
// directions. Heartbeats/election deliberately not wired yet - travel is
// announced, and the mac never resurrects a buddy that left (away rule).
final class Coordination {
    static let secret = "buddy-doorknob"
    static let serviceType = "_buddy._tcp"
    static let port: UInt16 = 47800

    let deviceId: String
    let rank: Int
    private(set) var epoch: Int
    private(set) var ownsBuddy: Bool

    // Called on main. arrived: buddy travels IN (payload = state blob).
    // departed: our travel was acked, buddy left.
    var onArrive: (([String: Any]) -> Void)?
    var onDepart: (() -> Void)?
    // Every replicated state snapshot from the current owner (traits sync).
    var onSnapshot: (([String: Any]) -> Void)?
    // A follower asked to change a trait while we own buddy.
    var onTraitSet: ((String, Double) -> Void)?

    private let queue = DispatchQueue(label: "buddy.coord")
    private var listener: NWListener?
    private var browser: NWBrowser?
    private var peers: [String: NWEndpoint] = [:]
    // Main-thread mirror of peer names, for UI (menu) reads.
    private(set) var knownPeers: [String] = []
    private var peerHosts: [String: String] = [:]
    private var connections: [NWConnection] = []
    private let stateURL: URL
    private var heartbeatTimer: DispatchSourceTimer?
    private var heartbeatSocket: NWListener?
    // Snapshot provider set by the controller - called on main, returns the
    // replicable state blob (traits etc). Broadcast to followers.
    var snapshot: (() -> [String: Any])?

    init(deviceId: String, rank: Int, owner: Bool, listenPort: UInt16 = Coordination.port) {
        self.deviceId = deviceId
        self.rank = rank
        self.ownsBuddy = owner
        self.stateURL = BuddyPaths.home.appendingPathComponent("coordination-\(deviceId).json")
        // Persisted state wins over the default: a mac that restarts after
        // buddy traveled away must not wake up thinking it still owns buddy.
        var e = 0
        var own = owner
        if let data = try? Data(contentsOf: stateURL),
           let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] {
            e = (json["epoch"] as? NSNumber)?.intValue ?? 0
            own = (json["owner"] as? Bool) ?? owner
        }
        self.epoch = e
        self.ownsBuddy = own
        start(listenPort: listenPort)
    }

    private func persist() {
        let json: [String: Any] = ["epoch": epoch, "owner": ownsBuddy]
        if let data = try? JSONSerialization.data(withJSONObject: json) {
            try? data.write(to: stateURL)
        }
    }

    // MARK: - Listen + discover

    private func start(listenPort: UInt16) {
        guard let l = try? NWListener(using: .tcp, on: NWEndpoint.Port(rawValue: listenPort)!) else {
            buddyLog("coord: cannot listen on \(listenPort)")
            return
        }
        l.service = NWListener.Service(name: deviceId, type: Coordination.serviceType)
        l.newConnectionHandler = { [weak self] conn in
            self?.adopt(conn)
        }
        l.start(queue: queue)
        listener = l

        let b = NWBrowser(for: .bonjour(type: Coordination.serviceType, domain: nil), using: .tcp)
        b.browseResultsChangedHandler = { [weak self] results, _ in
            guard let self else { return }
            var found: [String: NWEndpoint] = [:]
            for r in results {
                if case let .service(name, _, _, _) = r.endpoint, name != self.deviceId {
                    found[name] = r.endpoint
                }
            }
            let fresh = Set(found.keys).subtracting(self.peers.keys)
            self.peers = found
            // Deliberately NOT pruning peerHosts on browse loss: cached IPs
            // are the fallback for exactly when discovery goes blind.
            buddyLog("coord: peers \(Array(found.keys))")
            let names = Array(found.keys).sorted()
            DispatchQueue.main.async { self.knownPeers = names }
            for name in fresh { self.hello(name) }
        }
        b.start(queue: queue)
        browser = b

        startHeartbeats(listenPort: listenPort)
        buddyLog("coord: up as \(deviceId) rank \(rank) epoch \(epoch) owner \(ownsBuddy)")
    }

    // MARK: - Heartbeats + replication

    // UDP heartbeats to every peer every 3s (mesh, per coordination.md).
    // The mac takes NO action on peer silence: the phone going quiet means
    // "buddy is out with Pete", never death. Owner also piggybacks a state
    // snapshot broadcast every 10th beat so followers can resume after a crash.
    private func startHeartbeats(listenPort: UInt16) {
        let hbPort = listenPort + 1
        if let l = try? NWListener(using: .udp, on: NWEndpoint.Port(rawValue: hbPort)!) {
            l.newConnectionHandler = { [weak self] conn in
                guard let self else { return }
                conn.start(queue: self.queue)
                conn.receiveMessage { data, _, _, _ in
                    // Liveness intake only; nothing acts on it on the mac.
                    _ = data
                    conn.cancel()
                }
            }
            l.start(queue: queue)
            heartbeatSocket = l
        }
        var beat = 0
        let t = DispatchSource.makeTimerSource(queue: queue)
        t.schedule(deadline: .now() + 3, repeating: 3)
        t.setEventHandler { [weak self] in
            guard let self else { return }
            beat += 1
            for (_, host) in self.peerHosts {
                let conn = NWConnection(host: NWEndpoint.Host(host),
                                        port: NWEndpoint.Port(rawValue: hbPort)!,
                                        using: .udp)
                conn.start(queue: self.queue)
                var f: [String: Any] = ["type": "heartbeat", "epoch": self.epoch,
                                        "owner": self.ownsBuddy]
                f["v"] = 1
                f["s"] = Coordination.secret
                f["from"] = self.deviceId
                if let data = try? JSONSerialization.data(withJSONObject: f) {
                    conn.send(content: data, completion: .contentProcessed { _ in conn.cancel() })
                }
            }
            if beat % 10 == 0, self.ownsBuddy {
                self.broadcastState()
            }
        }
        t.resume()
        heartbeatTimer = t
    }

    // Full-snapshot state event to every peer (v1 replication: latest wins).
    // Owner-only: a follower pushing state would fight the owner's copy.
    func broadcastState() {
        DispatchQueue.main.async {
            guard self.ownsBuddy else { return }
            let payload = self.snapshot?() ?? [:]
            self.queue.async {
                for (_, endpoint) in self.peers {
                    let conn = NWConnection(to: endpoint, using: .tcp)
                    conn.start(queue: self.queue)
                    self.send(["type": "state", "epoch": self.epoch, "seq": 0,
                               "op": "snapshot", "owner": self.ownsBuddy,
                               "payload": payload], on: conn)
                    self.queue.asyncAfter(deadline: .now() + 2) { conn.cancel() }
                }
            }
        }
    }

    // Ask the current owner to change a trait (we are a follower). The
    // owner's broadcast echoes the clamped result back.
    func sendTraitSet(name: String, value: Double) {
        queue.async {
            guard !self.ownsBuddy, let (_, endpoint) = self.peers.first else { return }
            let conn = NWConnection(to: endpoint, using: .tcp)
            conn.start(queue: self.queue)
            self.send(["type": "traitSet", "epoch": self.epoch,
                       "name": name, "value": value], on: conn)
            self.queue.asyncAfter(deadline: .now() + 2) { conn.cancel() }
        }
    }

    // On meeting a peer: exchange hellos so a stale device learns the current
    // epoch before doing anything (cold-start grace + zombie correction).
    private func hello(_ name: String) {
        guard let endpoint = peers[name] else { return }
        let conn = NWConnection(to: endpoint, using: .tcp)
        conn.stateUpdateHandler = { [weak self] state in
            guard let self, case .ready = state else { return }
            // Remember the peer's raw IP: heartbeats are UDP to host:hbPort,
            // and cached IPs double as the discovery fallback.
            if case let .hostPort(host, _)? = conn.currentPath?.remoteEndpoint {
                self.peerHosts[name] = "\(host)".components(separatedBy: "%").first ?? "\(host)"
            }
        }
        adoptForFrames(conn)
        conn.start(queue: queue)
        send(["type": "hello", "epoch": epoch], on: conn)
    }

    private func adoptForFrames(_ conn: NWConnection) {
        connections.append(conn)
        receiveLines(conn, buffer: Data())
    }

    private func adopt(_ conn: NWConnection) {
        connections.append(conn)
        conn.start(queue: queue)
        receiveLines(conn, buffer: Data())
    }

    private func receiveLines(_ conn: NWConnection, buffer: Data) {
        var buf = buffer
        conn.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, done, err in
            guard let self else { return }
            if let data { buf.append(data) }
            while let nl = buf.firstIndex(of: 0x0A) {
                let line = buf[buf.startIndex..<nl]
                buf.removeSubrange(buf.startIndex...nl)
                if let frame = (try? JSONSerialization.jsonObject(with: line)) as? [String: Any] {
                    self.handle(frame, on: conn)
                }
            }
            if done || err != nil {
                self.connections.removeAll { $0 === conn }
                conn.cancel()
            } else {
                self.receiveLines(conn, buffer: buf)
            }
        }
    }

    private func send(_ frame: [String: Any], on conn: NWConnection) {
        var f = frame
        f["v"] = 1
        f["s"] = Coordination.secret
        f["from"] = deviceId
        guard var data = try? JSONSerialization.data(withJSONObject: f) else { return }
        data.append(0x0A)
        conn.send(content: data, completion: .contentProcessed { _ in })
    }

    // MARK: - Frames

    private func handle(_ frame: [String: Any], on conn: NWConnection) {
        guard frame["s"] as? String == Coordination.secret else { return }
        let fEpoch = (frame["epoch"] as? NSNumber)?.intValue ?? -1
        let type = frame["type"] as? String ?? ""

        // Stale-epoch rule (hello exempt: it's how stale peers learn).
        if type != "hello", fEpoch < epoch { return }

        switch type {
        case "hello":
            send(["type": "state", "epoch": epoch, "seq": 0,
                  "op": "snapshot", "owner": ownsBuddy], on: conn)

        case "travel":
            // Refuse deliveries addressed to someone else - a stale-cached IP
            // once looped a travel back to its sender (the self-delivery bug).
            if let to = frame["to"] as? String, to != deviceId {
                buddyLog("coord: travel misdelivery (to \(to)), refused")
                return
            }
            // Buddy incoming: adopt the proposed epoch, ack, own it.
            epoch = fEpoch
            ownsBuddy = true
            persist()
            send(["type": "travel-ack", "epoch": epoch], on: conn)
            let payload = frame["payload"] as? [String: Any] ?? [:]
            buddyLog("coord: travel in, epoch \(epoch)")
            DispatchQueue.main.async { self.onArrive?(payload) }

        case "traitSet":
            if ownsBuddy, let name = frame["name"] as? String,
               let value = (frame["value"] as? NSNumber)?.doubleValue {
                DispatchQueue.main.async { self.onTraitSet?(name, value) }
            }

        case "claim", "state":
            if type == "state", let payload = frame["payload"] as? [String: Any],
               frame["owner"] as? Bool == true, !ownsBuddy {
                DispatchQueue.main.async { self.onSnapshot?(payload) }
            }
            // Someone else claims/reports ownership. Higher epoch wins;
            // equal epoch resolves by rank (lower rank wins) - the split-brain
            // tiebreak from coordination.md.
            let theirRank = Coordination.rank(of: frame["from"] as? String ?? "")
            let theyOwn = (frame["owner"] as? Bool) ?? (type == "claim")
            guard theyOwn else { break }
            let theyWin = fEpoch > epoch || (fEpoch == epoch && theirRank < rank)
            if theyWin {
                epoch = fEpoch
                if ownsBuddy {
                    ownsBuddy = false
                    buddyLog("coord: ceding ownership to \(frame["from"] ?? "?") epoch \(fEpoch)")
                    DispatchQueue.main.async { self.onDepart?() }
                }
                persist()
            }

        default:
            break
        }
    }

    static func rank(of deviceId: String) -> Int {
        switch deviceId {
        case "mac": return 1
        case "phone": return 2
        default: return 9
        }
    }

    // MARK: - Travel out

    var hasPeer: Bool { !peers.isEmpty }

    // Best-effort travel to the first discovered peer. completion(main): true
    // if the target acked and buddy is gone from here.
    func travel(payload: [String: Any], completion: @escaping (Bool) -> Void) {
        queue.async {
            guard self.ownsBuddy, let (_, endpoint) = self.peers.first else {
                DispatchQueue.main.async { completion(false) }
                return
            }
            let proposed = self.epoch + 1
            let conn = NWConnection(to: endpoint, using: .tcp)
            var finished = false
            let finish: (Bool) -> Void = { ok in
                guard !finished else { return }
                finished = true
                if ok {
                    self.epoch = proposed
                    self.ownsBuddy = false
                    self.persist()
                    buddyLog("coord: travel out acked, epoch \(proposed)")
                    DispatchQueue.main.async { self.onDepart?() }
                } else {
                    conn.cancel()
                    buddyLog("coord: travel out failed/timeout")
                }
                DispatchQueue.main.async { completion(ok) }
            }
            self.adoptForAck(conn, expecting: proposed, finish: finish)
            conn.start(queue: self.queue)
            let target = self.peers.first?.0 ?? "phone"
            self.send(["type": "travel", "epoch": proposed, "to": target,
                       "payload": payload], on: conn)
            self.queue.asyncAfter(deadline: .now() + 10) { finish(false) }
        }
    }

    private func adoptForAck(_ conn: NWConnection, expecting: Int, finish: @escaping (Bool) -> Void) {
        connections.append(conn)
        func loop(_ buffer: Data) {
            var buf = buffer
            conn.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, done, err in
                guard let self else { return }
                if let data { buf.append(data) }
                while let nl = buf.firstIndex(of: 0x0A) {
                    let line = buf[buf.startIndex..<nl]
                    buf.removeSubrange(buf.startIndex...nl)
                    if let frame = (try? JSONSerialization.jsonObject(with: line)) as? [String: Any],
                       frame["s"] as? String == Coordination.secret,
                       frame["type"] as? String == "travel-ack",
                       (frame["epoch"] as? NSNumber)?.intValue == expecting {
                        finish(true)
                        return
                    }
                }
                if done || err != nil {
                    self.connections.removeAll { $0 === conn }
                } else {
                    loop(buf)
                }
            }
        }
        loop(Data())
    }
}
